pkg/objstore/objstore.go

```diff
@@ -172,29 +232,45 @@ func NopCloserWithSize(r io.Reader) io.ReadCloser {
 
 // UploadDir uploads all files in srcdir to the bucket with into a top-level directory
 // named dstdir. It is a caller responsibility to clean partial upload in case of failure.
-func UploadDir(ctx context.Context, logger log.Logger, bkt Bucket, srcdir, dstdir string) error {
+func UploadDir(ctx context.Context, logger log.Logger, bkt Bucket, srcdir, dstdir string, options ...UploadDirOption) error {
 	df, err := os.Stat(srcdir)
+	opts := applyUploadDirOptions(options...)
+	g, ctx := errgroup.WithContext(ctx)
+	guard := make(chan struct{}, opts.concurrency)
+
 	if err != nil {
 		return errors.Wrap(err, "stat dir")
 	}
 	if !df.IsDir() {
 		return errors.Errorf("%s is not a directory", srcdir)
 	}
-	return filepath.WalkDir(srcdir, func(src string, d fs.DirEntry, err error) error {
-		if err != nil {
-			return err
-		}
-		if d.IsDir() {
-			return nil
-		}
-		srcRel, err := filepath.Rel(srcdir, src)
-		if err != nil {
-			return errors.Wrap(err, "getting relative path")
-		}
+	err = filepath.WalkDir(srcdir, func(src string, d fs.DirEntry, err error) error {
+		guard <- struct{}{}
+		g.Go(func() error {
+			defer func() { <-guard }()
+			if err != nil {
+				return err
+			}
+			if d.IsDir() {
+				return nil
+			}
+			srcRel, err := filepath.Rel(srcdir, src)
+			if err != nil {
+				return errors.Wrap(err, "getting relative path")
+			}
+
+			dst := path.Join(dstdir, filepath.ToSlash(srcRel))
+			return UploadFile(ctx, logger, bkt, src, dst)
+		})
 
-		dst := path.Join(dstdir, filepath.ToSlash(srcRel))
-		return UploadFile(ctx, logger, bkt, src, dst)
+		return nil
 	})
+
+	if err == nil {
+		err = g.Wait()
```

Comment:
~~In any case, we should wait for all the other running goroutine or call them. We might leak goroutines here, no?~~

Reply:
NVM. `The first call to return a non-nil error cancels the group; its error will be returned by Wait.`

Reply:
Better to add a comment that error handling and context cancelling is handles by the `errgroup`

Reply:
👍